### PR TITLE
Integrate multiple problems into existing flow

### PR DESCRIPTION
**Pull request by alankbi**
### List of Changes:

* Improve UI for switching between problems
* Display which problem the player is currently viewing
* Disable looping when reaching first or last problem in the game
* Update UI to better support Windows interfaces
  * Since I'm not on Windows, @jasonz6688 could you run this branch locally and confirm whether these changes helped or not? Namely I added a backup font for the editor and attempted to disable the scroll bar for the leaderboard when no scroll is needed
* Dynamically update numProblems based on the number of selected problems
* Remove mixing of selected problems and random problems functionality on the backend
* Disable problem selection/removal actions on the frontend when loading to prevent concurrency bugs
* Improve existing setup for copyable content
  * Required boilerplate is cut down on and the copied message correctly disappears 2 seconds after the latest click

### Notes:

* The spectator page doesn't allow viewers to manually switch between problems, but I plan to implement this feature in the next PR (along with updating the leaderboard and results table setup)

### Screencast and Images:

Game page with problem navigation:
Game page with problem navigation

### CODE REVIEW

**Comment by alankbi** [comment]
[Integrate multiple problems into existing flow](https://trello.com/c/9SwagavL/406-integrate-multiple-problems-into-existing-flow)

**Comment by zpChris** [comment]
Really excited to review this, it all looks awesome! Going through #172 first, then I'll look at this 👍

**Comment by zpChris** [comment]
I have to head out a bit early today (I'm free all day tomorrow though), so wanted to provide some quicker feedback here:
1. When I ran the PR locally (using `localhost:8080`), I couldn't add more than one problem (it didn't automatically update the size for me?). Not sure why this might be, but that could be something to investigate.
2. I think the problem selection on the room page would be improved if it actually added them on the backend and loaded once you click "All good!", rather than on every new problem. Perhaps the problem selector doesn't even close when you click one problem, but that's not as important - I think having it add only on "All good!" would solve the delay problem and improve UX.

File: frontend/src/views/Results.tsx
Comment (zpChris) [inline frontend/src/views/Results.tsx:99]: nit: Remove this line if not used.

**Review by zpChris — COMMENTED** [review]
As always, this PR looks great! I left a few comments that mainly deal with potential improvements, but are not crucial to merging this PR - the only reason I didn't approve is because of my comments above on how the problem selection is not quite working correctly for me, for whatever reason.

I'll investigate a bit on my own setup to see what's happening locally, and comment here if I find anything. 👍 Excited to get this merged in!

**Comment by zpChris** [comment]
Upon more testing, it seems the error only happens with the admin account support@codejoust.co. I believe it may be fixed when just merging this PR with main to get the fixes introduced in #173 - if you end up doing that @alankbi, let me know if that works.

File: frontend/src/components/game/ProblemPanel.tsx
Comment (zpChris) [reply frontend/src/components/game/ProblemPanel.tsx:94]: I used the `webkit-line-clamp` setup on the problem cards at #175.

**Comment by alankbi** [comment]
> Upon further experimentation, I'm very confident this is the issue, it seems the frontend widget gets confused with the duplicate problems. 👍

Thanks for investigating this! Yeah, after merging it looks like the error doesn't exist anymore, so that should solve it

> I think the problem selection on the room page would be improved if it actually added them on the backend and loaded once you click "All good!", rather than on every new problem. Perhaps the problem selector doesn't even close when you click one problem, but that's not as important - I think having it add only on "All good!" would solve the delay problem and improve UX.

That's a good idea - implemented! It did require a decent amount of extra logic and I deleted the remove button on the normal 
 lobby page to make the modal the single source of truth for updating problems (and thus prevent any concurrency bugs hopefully). Feel free to test it out and lmk what you think (or if you notice any bugs lol)

File: frontend/src/components/game/ProblemPanel.tsx
Comment (alankbi) [reply frontend/src/components/game/ProblemPanel.tsx:94]: I think that looks great - done!

Comment (alankbi) [reply frontend/src/components/game/ProblemPanel.tsx:94]: This is a great idea - thanks for writing this all up :) Just implemented this change similar to how it's done in #175

File: frontend/src/components/game/ProblemPanel.tsx
Comment (alankbi) [reply frontend/src/components/game/ProblemPanel.tsx:103]: Hm yeah my main concern with this would be it causing a similar overlapping behavior as the console buttons, which can get weird if the screen size is small (or for the problem panel, if they simply scrolled down a little bit causing it to overlap with the description). Potentially we could add that row you mentioned to prevent it from overlapping, but I'm not sure if that's worth the design tradeoff/dev time to implement something like that - thoughts on leaving this for a post-MVP PR?

**Comment by alankbi** [comment]
Just re-requested review, although there are still some merge conflicts which I might wait to resolve until #172 is finalized, and the commit history seems to contain extraneous commits for some reason (possibly due to differences in merging between this and `stagingNumProblems`). I'll make sure those are all resolved before merging.

File: frontend/src/components/game/ProblemPanel.tsx
Comment (zpChris) [reply frontend/src/components/game/ProblemPanel.tsx:118]: Yes I think that's a valid concern - let's leave this for a potential post-MVP task!

File: frontend/src/components/special/CopyIndicator.tsx
Comment (zpChris) [reply frontend/src/components/special/CopyIndicator.tsx:110]: Actually out of curiosity, can you tell me how you implemented the two-second feature? I saw the code with the `useRef` but I'm not very familiar with that hook.

**Comment by zpChris** [comment]
> That's a good idea - implemented! It did require a decent amount of extra logic and I deleted the remove button on the normal lobby page to make the modal the single source of truth for updating problems (and thus prevent any concurrency bugs hopefully). Feel free to test it out and lmk what you think (or if you notice any bugs lol)

I love it! 💯 ❤️ I think this makes the process a _lot_ easier, especially on `localhost:3000` where the backend is slower to respond.

Slightly unrelated to this change but just general to the problem selector, I noticed that I was [unable to close the problem selector list](https://www.loom.com/share/7169a247b418401bb4fcc72e0278fbf9); I think this may have something to do with it being an element on the modal and not registering the location of other clicks, but I think it'd be helpful to have a "click outside" close that list.

There is also a slight note on whether selecting a problem should close that list or not - I'm split on this, so keep whatever you prefer.

**Review by zpChris — COMMENTED** [review]
Given the issue with all the additional commits added, there are 56 files changed, so I'm not going to go through the list; I'm commenting without explicit approval with some other nits.

1. Compilation with warnings: there are frontend style warnings that you may want to fix (screenshot below of the run after `npm start`).

npm start warnings

2. Socket Change Problem: This issue may be mute on your next PR, but I wanted to note that right now, on the spectator view, you can click the change problem buttons even though they don't do anything. (If this is irrelevant because of the next PR, don't worry about making a fix for this.)

3. Results Table Time: I think the Results table on the Spectator side shows "time" when it shouldn't - correct me if I'm wrong and it should be there.

4. Spectate Live Statistics: This may be another issue that you are fixing in the next PR, but I noticed that the spectator statistics in the live PR are not accurate; in the image below, I completely solved the first problem but not the second two, but the statistics still shows the score at 100% (even though I am on the second problem here). Whether it is scoring by problem or overall, this score is not accurate.

Spectator View Inaccurate Score

Outside of the above, this PR again looks great! Thanks for making the change on the problem title and back / forward buttons. 👍

**Comment by alankbi** [comment]
> 1. Compilation with warnings: there are frontend style warnings that you may want to fix (screenshot below of the run after `npm start`).

👍 

> 2. Socket Change Problem: This issue may be mute on your next PR, but I wanted to note that right now, on the spectator view, you can click the change problem buttons even though they don't do anything. (If this is irrelevant because of the next PR, don't worry about making a fix for this.)

> 4. Spectate Live Statistics: This may be another issue that you are fixing in the next PR, but I noticed that the spectator statistics in the live PR are not accurate; in the image below, I completely solved the first problem but not the second two, but the statistics still shows the score at 100% (even though I am on the second problem here). Whether it is scoring by problem or overall, this score is not accurate.

Yup, both of these are features left for my upcoming PR on the `NewLeaderboard` branch!

> 3. Results Table Time: I think the Results table on the Spectator side shows "time" when it shouldn't - correct me if I'm wrong and it should be there.

Is this referring to the time column? If so, I wasn't aware we wanted to remove that (it should still be the same best submission time overall or per problem as of the `NewLeaderboard` branch). Or lmk as well if you were referring to something else

> Slightly unrelated to this change but just general to the problem selector, I noticed that I was unable to close the problem selector list; I think this may have something to do with it being an element on the modal and not registering the location of other clicks, but I think it'd be helpful to have a "click outside" close that list.

Hm I'm also not able to reproduce this - do you have any steps for reproing? Currently running on localhost:8080 logged into the admin account

**Comment by alankbi** [comment]
> Hm I'm also not able to reproduce this - do you have any steps for reproing? Currently running on localhost:8080 logged into the admin account

Although I will note when closing/reopening the modal and then clicking immediately to the left/right of the open problem selector, sometimes the first click doesn't register; however, this only happens occasionally and only for the first click for me

File: frontend/src/components/special/CopyIndicator.tsx
Comment (alankbi) [reply frontend/src/components/special/CopyIndicator.tsx:110]: Basically setTimeout returns some kind of identifier/value that you can use to reference the timeout - using this, anytime the button is clicked, it will first check whether a previous timeout is already scheduled (with useRef) and cancel it if so before scheduling a new timeout. 

The reason I used useRef to store the returned identifiers was actually cause of something you mentioned haha - from what I understood in that convo it's similar to useState in terms of storing values but doesn't trigger a re-render when the value changes

**Comment by zpChris** [comment]
> Is this referring to the time column? If so, I wasn't aware we wanted to remove that (it should still be the same best submission time overall or per problem as of the NewLeaderboard branch). Or lmk as well if you were referring to something else

Yes, that's what I was referring to. I think it's inaccurate because if you are five minutes into the game, and click "Submit", it shows up as "5m ago" (when in reality, the submission occurred zero minutes ago but five minutes from the start of the game).

I made [this screencast](https://www.loom.com/share/c3d590c2e0034c4d800e3a4628a9db7c) to explain what I see as the issues in the game, spectator view, and results table - sorry if it's confusing as I was gathering my thoughts during the screencast, but we can talk more about this tonight.

> Hm I'm also not able to reproduce this - do you have any steps for reproing? Currently running on localhost:8080 logged into the admin account
> Although I will note when closing/reopening the modal and then clicking immediately to the left/right of the open problem selector, sometimes the first click doesn't register; however, this only happens occasionally and only for the first click for me

I don't think the issue of closing / reopening the modal is a big deal; I just tested on `localhost:8080` again right now and it worked great this time (who knows ¯\_(ツ)_/¯), but I found a different issue: as shown by [this screencast](https://www.loom.com/share/5e39e605bdb14a818a625fb701b70796), when you close the modal it does not trigger the save problems, and nothing happens when refreshing the page either. I think ideally, closing the modal should discard the changes, and refreshing the page should also refresh the problems selected (however, if you fix the first problem the second isn't a big issue).

Let me know if this all makes sense @alankbi, sorry I keep finding little nits haha.

**Comment by alankbi** [comment]
> I found a different issue: as shown by [this screencast](https://www.loom.com/share/5e39e605bdb14a818a625fb701b70796), when you close the modal it does not trigger the save problems, and nothing happens when refreshing the page either. I think ideally, closing the modal should discard the changes, and refreshing the page should also refresh the problems selected (however, if you fix the first problem the second isn't a big issue).

Good catch on this! I updated the close modal function to discard those changes if you exit without saving. 

> Yes, that's what I was referring to. I think it's inaccurate because if you are five minutes into the game, and click "Submit", it shows up as "5m ago" (when in reality, the submission occurred zero minutes ago but five minutes from the start of the game).
> 
> I made [this screencast](https://www.loom.com/share/c3d590c2e0034c4d800e3a4628a9db7c) to explain what I see as the issues in the game, spectator view, and results table - sorry if it's confusing as I was gathering my thoughts during the screencast, but we can talk more about this tonight.

I see, thanks for clearing that up! Yeah that makes more sense (time showing wrong rather than the time column being removed) - I will note that in the PR description, I mentioned the following comment:

> The spectator page doesn't allow viewers to manually switch between problems, but I plan to implement this feature in the next PR (along with updating the leaderboard and results table setup)

So I hadn't actually planned on making any changes related to the results/leaderboard in this PR but rather implement those in the next one (here's actually the [commit](https://github.com/CodeJoustHQ/main/commit/f626c7144daf8b96c6eccb577264b244f5636fff) that I made on that branch which fixes the mentioned time issue). If the concern is merging not-complete code into main, maybe one option is once both PRs are approved separately, we merge both in at once? Or if you prefer I can cherry-pick a few of the fixes into this PR before merging; lmk what you think

**Comment by zpChris** [comment]
> Good catch on this! I updated the close modal function to discard those changes if you exit without saving.

Thanks! 💯 

> I see, thanks for clearing that up! Yeah that makes more sense (time showing wrong rather than the time column being removed) - I will note that in the PR description, I mentioned the following comment:
> 
> > The spectator page doesn't allow viewers to manually switch between problems, but I plan to implement this feature in the next PR (along with updating the leaderboard and results table setup)
> 
> So I hadn't actually planned on making any changes related to the results/leaderboard in this PR but rather implement those in the next one (here's actually the [commit](https://github.com/CodeJoustHQ/main/commit/f626c7144daf8b96c6eccb577264b244f5636fff) that I made on that branch which fixes the mentioned time issue). If the concern is merging not-complete code into main, maybe one option is once both PRs are approved separately, we merge both in at once? Or if you prefer I can cherry-pick a few of the fixes into this PR before merging; lmk what you think

This all sounds good - I'm cool with you merging in this PR, and we can revisit these points on your next PR that you mentioned to double-check that everything is cleared up. Looks great!

File: frontend/src/views/Lobby.tsx
Comment (zpChris) [inline frontend/src/views/Lobby.tsx:461]: Should we set modified problems here, or is that not needed for some reason?

**Review by zpChris — APPROVED** [review]
LGTM! ✅

File: frontend/src/views/Lobby.tsx
Comment (alankbi) [reply frontend/src/views/Lobby.tsx:461]: It should be called in the `updateSelectedProblems` function once the POST request has gone through, I think it was done that way to prevent it from clearing the selection and showing the old problems while the request is loading

File: frontend/src/components/game/PlayerGameView.tsx
Comment (alankbi) [reply frontend/src/components/game/PlayerGameView.tsx:362]: Sure, I can change this to a TODO or explanatory comment - I'll do that on #178 actually, that way I won't have to re-run the builds until it passes again and also so #176 shows the correct commits sooner.

**Pull request merged**
